[Sally] 스프링 카페 4단계 - 로그인 구현

sally-ksh opened the pull request:
안녕하세요. 

- [X] 로그인과 로그아웃이 기능이 정상적으로 동작한다.
- [X] 현재 상태가 로그인 상태이면 상단 메뉴에서 “로그아웃”, “개인정보수정”이 표시되어야 한다.
- [X] 현재 상테가 로그인 상태가 아니라면 상단 메뉴에서 “로그인”, “회원가입”이 표시되어야 한다.

- [X] (선택) 개인정보 수정 기능 추가
- [X] 로그인한 사용자는 자신의 정보를 수정할 수 있어야 한다.
- [X] 이름, 이메일만 수정할 수 있으며, 사용자 아이디는 수정할 수 없다.
- [X] 비밀번호가 일치하는 경우에만 수정 가능하다.

- 추가 기능
 - 개인정보 수정시 비밀번호 입력 제한 횟수 초과하여 입력하면 제한시간을 두어 입력 시도를 제한합니다.
   - 버튼 비활성화와 입력칸은 입력하지 못하도록 변경 됩니다. 
 - 로그아웃 후 로그인을 통해 비밀번호 입력 시도시에도 제한시간 내에서는 입력 시도를 제한합니다.


![1](https://user-images.githubusercontent.com/96989782/158035783-7ae5a5de-aa69-4d0a-a02c-6c2462ac40f4.PNG)

![2](https://user-images.githubusercontent.com/96989782/158035786-83794638-3e64-44a2-82b8-c6641a3d7300.PNG)

![3](https://user-images.githubusercontent.com/96989782/158035790-fb443bef-4bae-4a3b-a1f6-b4d6b4d10767.PNG)

![4](https://user-images.githubusercontent.com/96989782/158035793-6d106be7-86c6-4d38-9afd-3bb6a76e37f9.PNG)


 배포는 여러 시도를 해보았는데요, fork 인게 연관이 있는지... 최신 상태라는 메시지만 뜨고 배포 결과를 확인하지 못하고 있습니다.
 이번에 배포도 에러가 많았고, 작업하면서도 여러 시도와 에러를 수정하면서 정리하기가 벅찼었습니다.
- 개인정보 수정시 비밀번호 제한 두기를 해보고 싶어서 시도했다가 생각보다 복잡해서 혼났습니다.
  - 실제로는 사용자 히스토리 등을 기록하거나 다른 변경 이력들이 남을거 같은데요. 이렇게 일시적으로 입력 제한 두는 경우에 그 횟수 까지 .DB 저장하는 게 좋지 않다고 생각을 두고서, UpdateDto 통해서 작업을 했습니다.
  - 실제로는 DB 설계와 로직, 객체들이 어떻게 분리되는지 궁금합니다. FE/BE 분리한 경우를 생가하면 횟수 등에 대한 정보가 저장되야 될 듯도 싶고 잘 모르겠습니다.
  - 구현에 재미를 두고 작업해보고자 시도한 작업이니 요구사항 벗어났더라도 이해 부탁드립니다.  😌 (


이번에 주석도 추가해보고 노력해봤는데, 주석이 과할지 잘 모르겠습니다.
다음부터 좀 더 잘 정리해서 pr 하도록 하겠습니다. 
감사합니다.

sally-ksh committed:
refactor: apply some reviews

- type casting of Repository interface
  - change primitive type of generic type
- StringValidator's function changed in package of utils

sally-ksh committed:
feat: 로그인 화면 및 로그인 요청 처리

- login.html 추가
- MvcConfig - URL과 html 연동
  - priority 안할시 get 요청 불가
- POST /do_login
- MvcConfig 와의 충돌 피하기 위해 do_login으로 url 변경
- 게시글 목록은 첫 페이지로 로그인 하지 않은 사용자에게도 공개
- 로그인은 메인 url에서 요청 및 응답
  MainController 에 UserRepository DI
  - 검증 로직이 Controller 상에서 일어나 화면을 리턴하기 떼문에 비즈니스 로직과는
    무관하다고 생각하여 UserRepository를 DI

sally-ksh committed:
refactor: 리뷰 반영 메모리DB 자료구조 변경

- Map -> List

sally-ksh committed:
refactor: 리뷰반영 - 메모리 DB 자료구조 변경

- User와 Article의 MemoryRepository의 자료구조를  Map -> List 로 변경
 - List 로 변경하면서 중복되는 메서드들을 JdbcTypeConvertor 로 옮겼습니다.
   - 가독성이 더 안좋아진거 같고, 객체의 필드값들을 일일이 꺼내어 전달하는게 고민입니다.

sally-ksh committed:
feat: 로그인 화면 및 로그아웃

- templates/main/login.html
  - 로그인/로그아웃시 화면에 보여주는 버튼이 달라진다.
- MainController /logout
  - 로그아웃시 세션정보를 지운다.
- 로그인 후 리다이렉트 후  URL 의 jsessionid 에러 제거
  - application.properties

sally-ksh committed:
feat: URL id아닌 UserId(닉네임) 노출로 변경

- id가 필요없을 수 있다고 하셔서, 생각해보고자 url 부분부터 변경했습니다.
  - 닉네임의 경우 가입한 사용자간 공개되어 사용되기에 id 보단 좋다고 생각했습니다.
  - 최근 이용하는 몇몇 사이트를 보니 사용자 정보가 url로 붙이지 않는 곳의 방법들이 궁금해 졌어요.

- 헤로쿠 배포 시도로 Profile, system.properties 추가
  - git 연동 후 최신 상태라를 메시지만 뜨고 아무 변화 없어서 배포에 어려움을 겪고 있습니다.

- SessionUser 클래스
  - Session에 뭐가 담길지 고민중입니다.

sally-ksh committed:
feat & refactor: SessionUser 필드 및 생성자 변경, 개인정보 수정 요청 세션 추가, sql 추가

- SessionUser 학습상 실습했던 내용 제거 후 UserId를 할당받는 UserName 필드만 남겼습니다.
- 개인정보 수정 요청 시 로그인한 사용자, 개인정보 수정 요청하는 계정과 로그인한 사용자 정보 일치 여부를 확인 햇습니다.
- 실행시 create 문 추가와 insert 진행되어 테스트 할 수 있도록 했습니다.

sally-ksh committed:
refactor: renaming

sally-ksh committed:
feat: 비밀번호 제한 회수 초과한 입력 시도시 재로그인 및 개인정보 수정 제한 로직

- 개인정보 수정시 3회 이상 시도하여 오류시 비밀번호 입력에 대한 10분의 제한 시간을 두게 했습니다.
  - 개인정보 수정시 입력하는 비밀번호의 남은 입력 횟수, 시간 제한 메시지를 전달합니다.
- 로그아웃 후 로그인 시도시에는 해당 userId에 대해 조회를 통해 10분이 지나지 않은 경우 시간 제한이 지속 되도록 했습니다.
  - 단, 로그인시 수차례의 비밀번호 입력에 대한 추가로직은 작업하지 않았습니다.

tmdgusya commented on src/main/java/com/kakao/cafe/main/MainController.java:
```diff
@@ -1,25 +1,67 @@
 	private Logger logger = LoggerFactory.getLogger(MainController.class);
 
-	public MainController(ArticleService articleService) {
+	public MainController(ArticleService articleService, UserRepository userRepository) {
```
`Controller` 에서 repository 를 주입받는게 맞는걸까요? 저는 잘 모르겠네요.

tmdgusya commented on src/main/java/com/kakao/cafe/main/MainController.java:
```diff
@@ -1,25 +1,67 @@
+			logger.info("login empty");
+			return "redirect:/login";
+		}
+		User user = getUser.get();
```
Controller 에 너무 큰 로직이 들어있습니다.
어떻게 하면 이 코드들을 Service 로 분리 해낼 수 있을지 생각해보세요.

tmdgusya commented on src/main/java/com/kakao/cafe/qna/domain/Article.java:
```diff
@@ -29,6 +31,16 @@ public Article(Long id, String writer, String title, String content, LocalDate w
 
+	public boolean hasId(Long id) {
+		return this.id == id;
+	}
```
hasId 보다는 equals 로 보이는데요?

tmdgusya commented on src/main/java/com/kakao/cafe/common/utils/sql/JdbcTypeConvertor.java:
```diff
@@ -0,0 +1,38 @@
+	/*
+		List로 자료구조 변경하면서 공통된 로직의 메서드가 중복되어 별도의 메서드로 꺼내어 작업해보고자 생각했습니다.
+	 */
+	public static int getDataIdx(Long entityIdx) {
+		return Math.toIntExact(entityIdx-1);
+	}
+
+	public static Long getNextId(int size) {
+		return size + 1L;
+	}
```
사실 단순히 generate value 가 목적이면 keyHolder 가 있긴한데 ㅎㅎ
그래도 공통적인 부분을 따로 뽑아내려 하신게 좋군요.

다만 Generic 은 지금 클래스에서는 의미가 없어보이는데요? Util 성 클래스 처럼 보입니다.

sally-ksh replied:
네, 감사합니다. 🥇

tmdgusya commented on src/main/java/com/kakao/cafe/qna/infra/MemoryArticleRepository.java:
```diff
@@ -19,48 +19,54 @@
-					entity.getContent(),
-					entity.getWritingDate()));
-			return id;
+		if (id != null && this.hasId(id)) {
+			data.set(getDataIdx(entity.getId()), entity);
+			logger.info("question db update : {}", entity.getId());
+			return entity;
```
ㅎㅎ 저는 여기서 null 체크가 필요한가 싶어요.

`this.hasId(id)` 로도 체크가 될거 같은데요 ㅎㅎ?

sally-ksh replied:
한번 더 보니 다르게 보이네요.
감사합니다 🥇

tmdgusya commented on src/main/java/com/kakao/cafe/qna/infra/MemoryArticleRepository.java:
```diff
@@ -19,48 +19,54 @@
-	private Long getNextId() {
-		return data.size()+1L;
+	private boolean hasId(Long id) {
+		Optional<Article> article = getArticleId(id);
```
`getArticleId` 인데 `Optional<Article>` 인가요?
만약 Id 로 찾아오는게 맞는거라면 `getArticleById` 로 바꿔주세요 ㅎㅎ

tmdgusya commented on src/main/java/com/kakao/cafe/qna/infra/MemoryArticleRepository.java:
```diff
@@ -19,48 +19,54 @@
+	private boolean hasId(Long id) {
+		Optional<Article> article = getArticleId(id);
+		if (article.isEmpty()) {
+			logger.error("not exist of article id : {}", id);
```
error 를 찍을땐 exception 도 넘겨주는게 좋아요.

tmdgusya replied:
한번 공부해보세요 ㅎㅎ

tmdgusya commented on src/main/java/com/kakao/cafe/user/domain/User.java:
```diff
@@ -1,28 +1,51 @@
+		this.restrictedEnterPassword = restrictedEnterPassword;
 	}
 
+	// 처음 가입시 사용되는 생성자
 	public User(String userId, String name, String email, String password) {
 		this.userId = userId;
 		this.name = name;
 		this.email = email;
 		this.password = password;
+		this.createdDate = LocalDateTime.now();
+		this.lastUpdatedDate = LocalDateTime.now();
+		this.restrictedEnterPassword = false;
 	}
```
이렇게 주석으로 생성자를 설명해야 한다면 어떻게 하는게 좋을까요?
저는 보통 생성자를 설명해야만 한다면 팩토리 패턴을 이용하는 편이에요.

tmdgusya commented on src/main/java/com/kakao/cafe/user/domain/User.java:
```diff
@@ -1,28 +1,51 @@
+	public boolean hasId(Long id) {
+		return this.id == id;
+	}
+	public boolean hasUserId(String userId) {
```
has 와 equal 의 차이를 공부하셔야 할거 같습니다.

tmdgusya commented on src/main/java/com/kakao/cafe/user/domain/User.java:
```diff
@@ -32,11 +55,50 @@ public boolean hasName(String name) {
+		}
+	}
+
+	private boolean isPassedLimitTime() {
+		return LocalDateTime.now().isAfter(this.lastUpdatedDate.plusMinutes(USER_LIMIT_TIME));
+	}
```
로그인할때 마다 업데이트 해주는건가요? ㅎㅎ 재밌네요.
미션외에 별개 사항이니 별도로 현업에서 방법에 대한 코멘트는 드리지는 않겠습니다.

tmdgusya commented on src/main/java/com/kakao/cafe/user/domain/UserController.java:
```diff
@@ -48,22 +57,52 @@ public String readProfile(@PathVariable Long id, Model model) {
+			비밀번호 입력 제한시 수정 요청시에는 수정요청이 불가하다는 별도의 안내 메시지를 전달한다.
+			비밀번호 입력 제한 10분이 지난 후에는 변경 요청이 가능하다.
+		 */
+		if (response.hasMessage()) {
+			model.addAttribute("notAllow", response.getMessage());
+		}
```
http status 로 표현은 안될까요 ㅎㅎ

tmdgusya commented on src/main/java/com/kakao/cafe/user/domain/UserController.java:
```diff
@@ -48,22 +57,52 @@ public String readProfile(@PathVariable Long id, Model model) {
 		return "redirect:/users/";
 	}
 
+	private void setTimeLimit(String userId, RedirectAttributes redirectAttributes,
+		WrongPasswordResponse passwordResponse) {
+		if (!passwordResponse.isValidChangingPassword()) {
+			userService.restrictPasswordChange(userId);
+			redirectAttributes.addFlashAttribute("notAllow", USER_MESSAGE_OF_EXCEED_PASSWORD_ENTRY);
+		}
+	}
```
사실 저는 로그인 제한이면 403 이니까 exception 을 throw 해줘야 한다고 생각하는데 ㅎㅎ.. 한번 고민해보세요.

sally-ksh replied:
호곡

tmdgusya commented on src/main/java/com/kakao/cafe/user/domain/UserService.java:
```diff
@@ -45,4 +48,40 @@ public void changeProfile(UserUpdateDto.Request userDto) {
+		return response;
+	}
+
+	private boolean isValidPasswordEntry(User user) {
+		if (user.isAllowedStatusOfPasswordEntry()) {
+			userRepository.save(user);
+			return true;
+		}
+		return false;
+	}
```
이게 정확히 어떤 로직인지 파악이 안되요 ㅎㅎ;;..

isValid 면 유효성 체크만 할거 같은데 내부에서 `userRepository.save(user);` 가 호출되네요.

sally-ksh replied:
3회 비밀번호 시도 초과 후에 DB의 restricted_enter_password = true 로 저장되게 했습니다.
로그아웃 하고서 로그인시 동일하게 패스워드 입력 시도를 막는다고 생각하고 작업했습니다.
DB 최대한 안 쓰려고 생각한 작업인데, 그러면 Dto 가 무거워지고 
좀더 리팩토링이나 변경은 구현 후에 하자고 생각하면서 작업 해봤습니다.
근데 에러가 많고 생각보다 단순하지 않아서 ... 리팩토링, 클린코드까진 어렵더라구요 ㅎㅎ
점점 구현 위한 작업에 몰두 했던 거 같습니다.

자세히 봐주셔서 감사합니다 👍

tmdgusya replied:
음 클린코드 / 리팩토링 이라기 보다는 유효성 검사를 하는 곳에서 `save` 와 같은 도메인 로직이 발생하는것이 맞는가에 대한 의문이였습니다.

sally-ksh replied:
아, 네, 메서드명 변경 하겠습니다. ^^

tmdgusya commented on src/main/java/com/kakao/cafe/user/domain/UserUpdateDto.java:
```diff
@@ -9,26 +9,71 @@
+	 */
+	public static class WrongPasswordResponse {
+		private static int LIMIT_NUMBER_OF_CHANGING_PASSWORD = 2;
+		private String message;
+		private int numberOfChangingPassword;
+		private boolean invalidPasswordEntry;
+
+		public static WrongPasswordResponse from(UserUpdateDto.Request userDto) {
+			WrongPasswordResponse response = new WrongPasswordResponse();
+			response.numberOfChangingPassword = userDto.getNumberOfChangingPassword();
+			response.message = getMessageAboutAttemptPassword(response);
+			response.invalidPasswordEntry = false;
+			return response;
+		}
+
+		public void restrictPasswordEntry() {
+			this.invalidPasswordEntry = true;
+		}
+
+		private static String getMessageAboutAttemptPassword(WrongPasswordResponse response) {
+			if (response.invalidPasswordEntry || response.isValidChangingPassword()) {
+				return String.format("비밀번호를 다시 입력하세요. %d회 남았습니다.",
+					response.getNumberOfPossibleChangingPassword());
+			} else {
+				return "비밀번호 오류제한 3회 초과 했습니다. 10분 후 다시 시도해 주세요.";
+			}
+		}
+
+		public int getNumberOfPossibleChangingPassword() {
+			return LIMIT_NUMBER_OF_CHANGING_PASSWORD - numberOfChangingPassword;
+		}
+
+		public boolean isValidChangingPassword() {
+			return this.numberOfChangingPassword < LIMIT_NUMBER_OF_CHANGING_PASSWORD;
+		}
+
+		public void addCount() {
+			this.numberOfChangingPassword++;
+		}
+	}
```
Response 를 만드신건 좋네요.
다만 단순히 Data Model 에 연산이 너무 많은건 아닐까? 라는 생각을 해봅니다
DTO 에 대해서 좀 더 공부를 해보셔야 할 거 같아요 ㅎㅎ

sally-ksh replied:
정말 고민이예여
이러면 안 될거 같은데, view 로 에러메시지는 보내줘야 하고,
상태값들을 도메인에 저장하자니 DB와 무관한 필드들이 많아져도 되는지 고민 됐습니다.
DB 저장에서 분리하고, view 와의 요청/응답으로 처리되니까 DTO 가 최선이라는 생각으로 했는데요.
메시지도 많아지고... 저도 좀더 개선시키면 좋겠는데 
아직은 아이디어가 없어요 ㅎㅎㅎ 어려워요... ㅎㅎㅎ ^^

어쨌든 DTO 가 이러면 안된다 라는걸 확신하게 돼서 좋습니다.
감사합니다. 👍

tmdgusya commented on src/main/java/com/kakao/cafe/user/infra/JdbcTemplateUserRepository.java:
```diff
@@ -42,6 +43,9 @@
 		NAME("name", ":name", ":name"),
 		EMAIL("email", ":email", ":email"),
 		PASSWORD("password", ":password", ":password"),
+		CREATED_DATE("created_date", ":created_date", ":createdDate"),
+		LAST_UPDATED_DATE("last_updated_date", ":last_updated_date", ":lastUpdatedDate"),
+		RESTRICTED_ENTER_PASSWORD("restricted_enter_password", ":restricted_enter_password", ":restrictedEnterPassword"),
```
굳이 Enum 으로 관리를 해야하나? 라는 생각이 있긴해요 ㅎㅎ

tmdgusya replied:
그래도 만약 계속해서 Query 를 직접 작성해야 하는 상황이라면 좋은 선택일거 같습니다.

sally-ksh replied:
ㅎㅎㅎ 감사합니다. 
왠지 지우기가 쉽지 않습니다.
생각보다 쿼리 대신 사용은 편리해서 사용하게 되고 있지만,
미션 3에서 작업한 내용이구요.

다음 미션에서 새 테이블 생성시에는 학습상 쿼리로 작업하고자 합니다.

감사합니다.

tmdgusya commented on src/main/java/com/kakao/cafe/user/infra/MemoryUserRepository.java:
```diff
@@ -1,77 +1,112 @@
-					entity.getEmail(),
-					entity.getPassword()));
-			return id;
+		if (id != null && this.hasId(id)) {
+			data.set(getDataIdx(entity), entity);
+			logger.info("update user : {}", entity.getUserId());
+			return entity;
 		}
```
이 logic 무언가 봤던거 같은데 ㅎㅎ

sally-ksh replied:
네, 고치겠습니다. ^^

tmdgusya reviewed (changes requested):
4단계 미션 하시느라 고생많으셨습니다 👍 

일단 기본적으로 코드의 정리가 필요해 보입니다. 함수의 Naming 의 개선이 필요해보여요. 아니면 주석이라도요. 전반적으로 몇몇 함수들은 읽었을때만 보고는 그 함수가 어떤 역할을 하는지 정확히 알기 힘든 것들이 있었어요. 이런경우 주석을 사용해도 좋으니 주석을 사용해주시면 좋겠습니다.

새로운 구현을 해보려는 방법은 좋습니다 👍  ㅎㅎ. 재미를 느끼면서 코딩하시는 것 또한 중요하니까요.
다만 이번 미션4 가 HTTP 에 대한 내용을 다루는 만큼, 시간이 남는다면 해당 부분으로 더 잘 표현할 수 있는 방법을 찾아보고 그 이후에 시간이 남는다면 새로운 것들을 시도해보셨으면 어땠을까? 라는 생각이 들어요. 이렇게 되면 미션에서 얻어갈 수 있는 것들을 얻어가기 어려울 수도 있다고 생각이 들어서요.

Enum 등등 코드를 한곳으로 모아 관리하려는 노력 좋습니다~!
Service, Controller, Repository 는 어떤 역할일까요? 왜 Controller 에서 최대한 Repsitory 의 사용을 피해야 할까요 ㅎㅎ. 사용하신만큼 고민해보세요.

sally-ksh committed:
refactor: 리뷰 반영
>>
>> - hasId : 메서드명에 맞는 로직 정정
>> - Dto 수정
>> - Controller 레이어에서 Repository 분리
>> - isValidPasswordEntry() -> checkStatusOfPasswordEntry()
>>   - 유효성 검사가 아닌, 상태체크에 따른 update 로직

sally-ksh committed:
refactor: 에러 로그 정정

- 로그 에러에 exception도 함께 넘기도록 변경했습니다.

sally-ksh commented:
수정하면서 새로운 점들을 많이 알게 되었고,
새로 눈에 띄이는 것들도 있어서 좋았습니다.

고친 코드도 어색한 거 같네요.. 좀 더 나아지면 좋겠습니다 ㅎㅎ

좋은 리뷰와 지시 감사합니다. 👍 🥇

tmdgusya commented:
좋습니다 ㅎㅎ
